### PR TITLE
feat: remote input subscribe on barrier mutation

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Currently, `RemoteInput` fetches the mutation of each barrier from `LocalBarrierWorker` by per-barrier sending a request to it and wait on a response oneshot channel.

This mechanism has some disadvantages:
1. when the number of remote input increases, the `LocalBarrierWorker` may get a great number of these request, which may block the handling of other request.
2. The remote input will always be blocked when receiving a barrier.

In this PR, we change to using a long existing channel to subscribe on barrier mutation. On the first barrier, the `RemoteInput` will send a `SubscribeBarrierMutation` request to the `LocalBarrierWorker`, including the actor_id and start epoch. When receiving this request, the `LocalBarrierWorker` will respond with the mutation of barriers that are already injected and has epoch later than start epoch. When later receiving barrier, it will push the mutation to the sender so that the `RemoteInput` can receive the mutation.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by fuyufjh** [comment]
> 2\. The remote input will always be blocked when receiving a barrier.

Didn't get this point. I thought it's necessary for the `RemoteInput` to block and wait for the Mutation barrier? How does this PR address this problem?

**Comment by wenym1** [comment]
> Didn't get this point. I thought it's necessary for the `RemoteInput` to block and wait for the Mutation barrier? How does this PR address this problem?

For the usual case that `LocalBarrierWorker` receivers the barrier earlier than the `RemoteInput`. In current main branch, the `RemoteInput` sends a request and waits for the response, which must be blocking (poll return `Pending`). With this PR, the `LocalBarrierWorker` will send the mutation to the channel when receiving the barrier, and when `RemoteInput` later receives the barrier, it is very likely that it won't be blocked (poll return `Ready`) because the mutation is already in the channel.

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 1]
```diff
@@ -193,12 +182,27 @@ impl Display for ManagedBarrierStateDebugInfo<'_> {
     }
 }
 
+#[derive(Default)]
+struct ActorMutationSubscribers {
+    #[expect(clippy::type_complexity)]
+    pending_subscribers: BTreeMap<u64, Vec<mpsc::UnboundedSender<(u64, Option<Arc<Mutation>>)>>>,
+    started_subscribers: Vec<mpsc::UnboundedSender<(u64, Option<Arc<Mutation>>)>>,
```
Comment (fuyufjh) [inline src/stream/src/task/barrier_manager/managed_state.rs:189]: The type complexity here indeed looks too high... How about

```suggestion
    pending_subscribers: BTreeMap<u64, Vec<MutationSubscriber>>,
    started_subscribers: Vec<MutationSubscriber>,
```

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 2]
```diff
@@ -193,12 +182,27 @@ impl Display for ManagedBarrierStateDebugInfo<'_> {
...
+    }
+}
+
 pub(super) struct ManagedBarrierState {
     /// Record barrier state for each epoch of concurrent checkpoints.
     ///
     /// The key is `prev_epoch`, and the first value is `curr_epoch`
     epoch_barrier_state_map: BTreeMap<u64, BarrierState>,
 
+    mutation_subscrbers: HashMap<ActorId, ActorMutationSubscribers>,
```
Comment (fuyufjh) [inline src/stream/src/task/barrier_manager/managed_state.rs:204]: Typo 

```suggestion
    mutation_subscribers: HashMap<ActorId, ActorMutationSubscribers>,
```

File: src/stream/src/executor/exchange/input.rs [hunk 3]
```diff
@@ -203,10 +205,29 @@ impl RemoteInput {
...
+                            let mutation = mutation_subscriber
+                                .recv()
                                 .await
-                                .context("Read barrier mutation error")?;
+                                .ok_or_else(|| {
+                                    anyhow!("failed to receive mutation of barrier {:?}", barrier)
+                                })
+                                .and_then(|(prev_epoch, mutation)| {
+                                    if prev_epoch != barrier.epoch.prev {
+                                        Err(anyhow!(
```
Comment (BugenZhao) [inline src/stream/src/executor/exchange/input.rs:222]: It seems to be a bug if this happens. What about `unreachable!`?

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 4]
```diff
@@ -69,13 +68,6 @@ impl Debug for IssuedState {
 /// The state machine of local barrier manager.
 #[derive(Debug)]
 enum ManagedBarrierStateInner {
-    /// Received barrier from actors in other compute nodes in remote input, however no `send_barrier`
-    /// request from the meta service is issued.
-    Stashed {
-        /// Senders registered by the remote input.
-        mutation_senders: Vec<oneshot::Sender<Option<Arc<Mutation>>>>,
-    },
-
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:78]: It's quite clever that the new impl eliminates the need for the `Stashed` state.

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 5]
```diff
@@ -246,39 +251,69 @@ impl ManagedBarrierState {
...
+                                    if tx
+                                        .send((*prev_epoch, issued_state.mutation.clone()))
+                                        .is_err()
+                                    {
+                                        // No more subscribe on the mutation. Simply return.
+                                        return;
+                                    }
+                                } else {
+                                    // The barrier no more collect from such actor. End subscribe on mutation.
+                                    return;
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:280]: Is this actually unreachable?

**Review by BugenZhao — APPROVED** [review]
Rest LGTM

**Review by fuyufjh — COMMENTED** [review]
The rest LGTM

File: src/stream/src/executor/exchange/input.rs (hunk 3 above)
Comment (yezizp2012) [reply src/stream/src/executor/exchange/input.rs:222]: Or add assertion for it?

**Review by yezizp2012 — APPROVED** [review]
LGTM

**Pull request merged**